### ISSUE TITLE
When server closes TCP connection because TCP window is full, connection recovery does not kick in

### ISSUE THREAD

**Issue by vincentjames501** [rabbitmq/rabbitmq-java-client#341]
## Summary

We started seeing the following in production:

```
java.net.SocketException: Broken pipe
	at java.net.SocketOutputStream.socketWrite0(Native Method)
	at java.net.SocketOutputStream.socketWrite(SocketOutputStream.java:109)
	at java.net.SocketOutputStream.write(SocketOutputStream.java:153)
	at java.io.BufferedOutputStream.flushBuffer(BufferedOutputStream.java:82)
	at java.io.BufferedOutputStream.flush(BufferedOutputStream.java:140)
	at java.io.DataOutputStream.flush(DataOutputStream.java:123)
	at com.rabbitmq.client.impl.SocketFrameHandler.flush(SocketFrameHandler.java:177)
	at com.rabbitmq.client.impl.AMQConnection.flush(AMQConnection.java:559)
	at com.rabbitmq.client.impl.AMQCommand.transmit(AMQCommand.java:127)
	at com.rabbitmq.client.impl.AMQChannel.quiescingTransmit(AMQChannel.java:396)
	at com.rabbitmq.client.impl.AMQChannel.transmit(AMQChannel.java:372)
	at com.rabbitmq.client.impl.AMQChannel.transmit(AMQChannel.java:365)
	at com.rabbitmq.client.impl.ChannelN.basicAck(ChannelN.java:1169)
	at com.rabbitmq.client.impl.recovery.RecoveryAwareChannelN.basicAck(RecoveryAwareChannelN.java:89)
	at com.rabbitmq.client.impl.recovery.AutorecoveringChannel.basicAck(AutorecoveringChannel.java:436)
	at RMQProblemTest$1.handleDelivery(RMQProblemTest.java:87)
	at com.rabbitmq.client.impl.ConsumerDispatcher$5.run(ConsumerDispatcher.java:149)
	at com.rabbitmq.client.impl.ConsumerWorkService$WorkPoolRunnable.run(ConsumerWorkService.java:100)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
```

After looking at a packet capture, we were seeing `[TCP Window Full]` messages when RabbitMQ was sending to our consumer followed by a ton of `[TCP ZeroWindow]` frames coming from the consumer. After enough of these, RabbitMQ abruptly closes the connection by sending a `RST` frame and reporting only this in the logs:

```
=ERROR REPORT==== 10-Jan-2018::20:38:08 ===
closing AMQP connection <0.5373.1> ([::1]:53814 -> [::1]:5672):
{writer,send_failed,{error,timeout}}
```

Further investigation revealed that I was setting the QoS for the consumer only *AFTER* I had already started consuming (i.e. I was calling `basicConsume` directly before `basicQos`). I realize that this was wrong of me, however, it seems odd that this would cause the exception above especially with such a relatively small number of small messages. The more concerning thing is that none of the recovery/shutdown methods on either the connection or channel seemed to be called despite the fact that the connection was indeed closed.

## Reproduction

I made a little test project that shows the issue.

https://github.com/vincentjames501/rmq-fail/blob/master/src/test/java/RMQProblemTest.java

Running the test will show the issue after 30 seconds or so. It appears reproducible on most RabbitMQ versions but I documented by setup below.

## Environment

RabbitMQ server - *3.6.12*
Erlang - *Erlang 20.1*
Operating system version (and distribution, if applicable) - *OSX 10.13.2*
All client libraries used - *RabbitMQ Java Client 4.4.1*
RabbitMQ plugins (if applicable) - *Just Management Console*

### Thanks!

**Comment by michaelklishin** [comment]
Thank you for your time.

Team RabbitMQ uses GitHub issues for specific actionable items engineers can work on. This assumes two things:

1. GitHub issues are not used for questions, investigations, root cause analysis, discussions of potential issues, etc (as defined by this team)
2. We have a certain amount of information to work with

We get at least a dozen of questions through various venues every single day, often quite light on details.
At that rate GitHub issues can very quickly turn into a something impossible to navigate and make sense of even for our team. Because of that questions, investigations, root cause analysis, discussions of potential features are all considered to be mailing list material by our team. Please post this to [rabbitmq-users](https://groups.google.com/forum/#!forum/rabbitmq-users).

Getting all the details necessary to reproduce an issue, make a conclusion or even form a hypothesis about what's happening can take a fair amount of time. Our team is multiple orders of magnitude smaller than the RabbitMQ community. Please help others help you by providing a way to reproduce the behavior you're
observing, or at least sharing as much relevant information as possible [on the list](https://groups.google.com/forum/#!forum/rabbitmq-users):

 * Server, client library and plugin (if applicable) versions used
 * Server logs
 * A code example or terminal transcript that can be used to reproduce
 * **Full** exception stack traces (not a single line message)
 * `rabbitmqctl status` (and, if possible, `rabbitmqctl environment` output)
 * Other relevant things about the environment and workload, e.g. [a traffic capture](https://www.rabbitmq.com/amqp-wireshark.html)

Feel free to edit out hostnames and other potentially sensitive information.

When/if we have enough details and evidence we'd be happy to file a new issue.

Thank you.

**Comment by michaelklishin** [comment]
The message means that RabbitMQ tried to write data to a socket and that failed with a timeout. See [Heartbeats](http://www.rabbitmq.com/heartbeats.html) and [Troubleshooting Network Connectivity](http://www.rabbitmq.com/troubleshooting-networking.html). There is nothing RabbitMQ or this client can do to completely avoid this kind of scenarios, so using [automatic connection recovery](http://www.rabbitmq.com/api-guide.html) is highly recommended.

**Comment by michaelklishin** [comment]
I don't see how this `java.net.SocketException` would be different from other socket exceptions or why that would not trigger connection recovery. Maybe @acogoluegnes recalls anything similar reported. Thanks for the details and the sample project that can help reproduce but this is mailing list material at this point.

**Comment by michaelklishin** [comment]
I executed the provided test. It simulates a slow consumer which does eventually lead to a full TCP buffer on the server end that trips up socket writes. It works well over localhost. Connection recovery does kick in, however: the "Consumed so far …" counter does keep growing:

```
/Library/Java/JavaVirtualMachines/jdk1.8.0_102.jdk/Contents/Home/bin/java -ea -Didea.launcher.port=7532 "-Didea.launcher.bin.path=/Applications/IntelliJ IDEA CE.app/Contents/bin" -Dfile.encoding=UTF-8 -classpath "/Applications/IntelliJ IDEA CE.app/Contents/lib/idea_rt.jar:/Applications/IntelliJ IDEA CE.app/Contents/plugins/junit/lib/junit-rt.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_102.jdk/Contents/Home/jre/lib/charsets.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_102.jdk/Contents/Home/jre/lib/deploy.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_102.jdk/Contents/Home/jre/lib/ext/cldrdata.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_102.jdk/Contents/Home/jre/lib/ext/dnsns.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_102.jdk/Contents/Home/jre/lib/ext/jaccess.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_102.jdk/Contents/Home/jre/lib/ext/jfxrt.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_102.jdk/Contents/Home/jre/lib/ext/localedata.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_102.jdk/Contents/Home/jre/lib/ext/nashorn.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_102.jdk/Contents/Home/jre/lib/ext/sunec.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_102.jdk/Contents/Home/jre/lib/ext/sunjce_provider.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_102.jdk/Contents/Home/jre/lib/ext/sunpkcs11.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_102.jdk/Contents/Home/jre/lib/ext/zipfs.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_102.jdk/Contents/Home/jre/lib/javaws.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_102.jdk/Contents/Home/jre/lib/jce.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_102.jdk/Contents/Home/jre/lib/jfr.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_102.jdk/Contents/Home/jre/lib/jfxswt.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_102.jdk/Contents/Home/jre/lib/jsse.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_102.jdk/Contents/Home/jre/lib/management-agent.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_102.jdk/Contents/Home/jre/lib/plugin.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_102.jdk/Contents/Home/jre/lib/resources.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_102.jdk/Contents/Home/jre/lib/rt.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_102.jdk/Contents/Home/lib/ant-javafx.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_102.jdk/Contents/Home/lib/dt.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_102.jdk/Contents/Home/lib/javafx-mx.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_102.jdk/Contents/Home/lib/jconsole.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_102.jdk/Contents/Home/lib/packager.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_102.jdk/Contents/Home/lib/sa-jdi.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_102.jdk/Contents/Home/lib/tools.jar:/private/tmp/rmq-fail/target/test-classes:/Users/antares/.m2/repository/com/rabbitmq/amqp-client/4.4.1/amqp-client-4.4.1.jar:/Users/antares/.m2/repository/org/slf4j/slf4j-api/1.7.21/slf4j-api-1.7.21.jar:/Users/antares/.m2/repository/junit/junit/4.12/junit-4.12.jar:/Users/antares/.m2/repository/org/hamcrest/hamcrest-core/1.3/hamcrest-core-1.3.jar" com.intellij.rt.execution.application.AppMain com.intellij.rt.execution.junit.JUnitStarter -ideVersion5 RMQProblemTest
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
Consumed so far - 0
Consumed so far - 0
Consumed so far - 0
Consumed so far - 0
Consumed so far - 1
Consumed so far - 1
Consumed so far - 2
Consumed so far - 2
Consumed so far - 2
Consumed so far - 3
Consumed so far - 3
Consumed so far - 3
Consumed so far - 4
Consumed so far - 4
Consumed so far - 4
Consumed so far - 5
Consumed so far - 5
Consumed so far - 5
Consumed so far - 6
Consumed so far - 6
Consumed so far - 6
Consumed so far - 7
Consumed so far - 7
Consumed so far - 7
Consumed so far - 8
Consumed so far - 8
Consumed so far - 8
Consumed so far - 9
Consumed so far - 9
Consumed so far - 9
Consumed so far - 10
Consumed so far - 10
Consumed so far - 10
Consumed so far - 11
Consumed so far - 11
java.net.SocketException: Broken pipe
	at java.net.SocketOutputStream.socketWrite0(Native Method)
	at java.net.SocketOutputStream.socketWrite(SocketOutputStream.java:109)
	at java.net.SocketOutputStream.write(SocketOutputStream.java:153)
	at java.io.BufferedOutputStream.flushBuffer(BufferedOutputStream.java:82)
	at java.io.BufferedOutputStream.flush(BufferedOutputStream.java:140)
	at java.io.DataOutputStream.flush(DataOutputStream.java:123)
	at com.rabbitmq.client.impl.SocketFrameHandler.flush(SocketFrameHandler.java:177)
	at com.rabbitmq.client.impl.AMQConnection.flush(AMQConnection.java:559)
	at com.rabbitmq.client.impl.AMQCommand.transmit(AMQCommand.java:127)
	at com.rabbitmq.client.impl.AMQChannel.quiescingTransmit(AMQChannel.java:396)
	at com.rabbitmq.client.impl.AMQChannel.transmit(AMQChannel.java:372)
	at com.rabbitmq.client.impl.AMQChannel.transmit(AMQChannel.java:365)
	at com.rabbitmq.client.impl.ChannelN.basicAck(ChannelN.java:1169)
	at com.rabbitmq.client.impl.recovery.RecoveryAwareChannelN.basicAck(RecoveryAwareChannelN.java:89)
	at com.rabbitmq.client.impl.recovery.AutorecoveringChannel.basicAck(AutorecoveringChannel.java:436)
	at RMQProblemTest$1.handleDelivery(RMQProblemTest.java:94)
	at com.rabbitmq.client.impl.ConsumerDispatcher$5.run(ConsumerDispatcher.java:149)
	at com.rabbitmq.client.impl.ConsumerWorkService$WorkPoolRunnable.run(ConsumerWorkService.java:100)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Consumed so far - 11
Consumed so far - 12
Consumed so far - 12
java.net.SocketException: Broken pipe
	at java.net.SocketOutputStream.socketWrite0(Native Method)
	at java.net.SocketOutputStream.socketWrite(SocketOutputStream.java:109)
	at java.net.SocketOutputStream.write(SocketOutputStream.java:153)
	at java.io.BufferedOutputStream.flushBuffer(BufferedOutputStream.java:82)
	at java.io.BufferedOutputStream.flush(BufferedOutputStream.java:140)
	at java.io.DataOutputStream.flush(DataOutputStream.java:123)
	at com.rabbitmq.client.impl.SocketFrameHandler.flush(SocketFrameHandler.java:177)
	at com.rabbitmq.client.impl.AMQConnection.flush(AMQConnection.java:559)
	at com.rabbitmq.client.impl.AMQCommand.transmit(AMQCommand.java:127)
	at com.rabbitmq.client.impl.AMQChannel.quiescingTransmit(AMQChannel.java:396)
	at com.rabbitmq.client.impl.AMQChannel.transmit(AMQChannel.java:372)
	at com.rabbitmq.client.impl.AMQChannel.transmit(AMQChannel.java:365)
	at com.rabbitmq.client.impl.ChannelN.basicAck(ChannelN.java:1169)
	at com.rabbitmq.client.impl.recovery.RecoveryAwareChannelN.basicAck(RecoveryAwareChannelN.java:89)
	at com.rabbitmq.client.impl.recovery.AutorecoveringChannel.basicAck(AutorecoveringChannel.java:436)
	at RMQProblemTest$1.handleDelivery(RMQProblemTest.java:94)
	at com.rabbitmq.client.impl.ConsumerDispatcher$5.run(ConsumerDispatcher.java:149)
	at com.rabbitmq.client.impl.ConsumerWorkService$WorkPoolRunnable.run(ConsumerWorkService.java:100)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Consumed so far - 12
Consumed so far - 13

[edited out for brevity]

Consumed so far - 30
Consumed so far - 31
Consumed so far - 31
java.net.SocketException: Broken pipe
	at java.net.SocketOutputStream.socketWrite0(Native Method)
	at java.net.SocketOutputStream.socketWrite(SocketOutputStream.java:109)
	at java.net.SocketOutputStream.write(SocketOutputStream.java:153)
	at java.io.BufferedOutputStream.flushBuffer(BufferedOutputStream.java:82)
	at java.io.BufferedOutputStream.flush(BufferedOutputStream.java:140)
	at java.io.DataOutputStream.flush(DataOutputStream.java:123)
	at com.rabbitmq.client.impl.SocketFrameHandler.flush(SocketFrameHandler.java:177)
	at com.rabbitmq.client.impl.AMQConnection.flush(AMQConnection.java:559)
	at com.rabbitmq.client.impl.AMQCommand.transmit(AMQCommand.java:127)
	at com.rabbitmq.client.impl.AMQChannel.quiescingTransmit(AMQChannel.java:396)
	at com.rabbitmq.client.impl.AMQChannel.transmit(AMQChannel.java:372)
	at com.rabbitmq.client.impl.AMQChannel.transmit(AMQChannel.java:365)
	at com.rabbitmq.client.impl.ChannelN.basicAck(ChannelN.java:1169)
	at com.rabbitmq.client.impl.recovery.RecoveryAwareChannelN.basicAck(RecoveryAwareChannelN.java:89)
	at com.rabbitmq.client.impl.recovery.AutorecoveringChannel.basicAck(AutorecoveringChannel.java:436)
	at RMQProblemTest$1.handleDelivery(RMQProblemTest.java:94)
	at com.rabbitmq.client.impl.ConsumerDispatcher$5.run(ConsumerDispatcher.java:149)
	at com.rabbitmq.client.impl.ConsumerWorkService$WorkPoolRunnable.run(ConsumerWorkService.java:100)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Consumed so far - 31
Consumed so far - 32
Consumed so far - 32
java.net.SocketException: Broken pipe
	at java.net.SocketOutputStream.socketWrite0(Native Method)
	at java.net.SocketOutputStream.socketWrite(SocketOutputStream.java:109)
	at java.net.SocketOutputStream.write(SocketOutputStream.java:153)
	at java.io.BufferedOutputStream.flushBuffer(BufferedOutputStream.java:82)
	at java.io.BufferedOutputStream.flush(BufferedOutputStream.java:140)
	at java.io.DataOutputStream.flush(DataOutputStream.java:123)
	at com.rabbitmq.client.impl.SocketFrameHandler.flush(SocketFrameHandler.java:177)
	at com.rabbitmq.client.impl.AMQConnection.flush(AMQConnection.java:559)
	at com.rabbitmq.client.impl.AMQCommand.transmit(AMQCommand.java:127)
	at com.rabbitmq.client.impl.AMQChannel.quiescingTransmit(AMQChannel.java:396)
	at com.rabbitmq.client.impl.AMQChannel.transmit(AMQChannel.java:372)
	at com.rabbitmq.client.impl.AMQChannel.transmit(AMQChannel.java:365)
	at com.rabbitmq.client.impl.ChannelN.basicAck(ChannelN.java:1169)
	at com.rabbitmq.client.impl.recovery.RecoveryAwareChannelN.basicAck(RecoveryAwareChannelN.java:89)
	at com.rabbitmq.client.impl.recovery.AutorecoveringChannel.basicAck(AutorecoveringChannel.java:436)
	at RMQProblemTest$1.handleDelivery(RMQProblemTest.java:94)
	at com.rabbitmq.client.impl.ConsumerDispatcher$5.run(ConsumerDispatcher.java:149)
	at com.rabbitmq.client.impl.ConsumerWorkService$WorkPoolRunnable.run(ConsumerWorkService.java:100)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Consumed so far - 32
Consumed so far - 33
Consumed so far - 33
java.net.SocketException: Broken pipe
	at java.net.SocketOutputStream.socketWrite0(Native Method)
	at java.net.SocketOutputStream.socketWrite(SocketOutputStream.java:109)
	at java.net.SocketOutputStream.write(SocketOutputStream.java:153)
	at java.io.BufferedOutputStream.flushBuffer(BufferedOutputStream.java:82)
	at java.io.BufferedOutputStream.flush(BufferedOutputStream.java:140)
	at java.io.DataOutputStream.flush(DataOutputStream.java:123)
	at com.rabbitmq.client.impl.SocketFrameHandler.flush(SocketFrameHandler.java:177)
	at com.rabbitmq.client.impl.AMQConnection.flush(AMQConnection.java:559)
	at com.rabbitmq.client.impl.AMQCommand.transmit(AMQCommand.java:127)
	at com.rabbitmq.client.impl.AMQChannel.quiescingTransmit(AMQChannel.java:396)
	at com.rabbitmq.client.impl.AMQChannel.transmit(AMQChannel.java:372)
	at com.rabbitmq.client.impl.AMQChannel.transmit(AMQChannel.java:365)
	at com.rabbitmq.client.impl.ChannelN.basicAck(ChannelN.java:1169)
	at com.rabbitmq.client.impl.recovery.RecoveryAwareChannelN.basicAck(RecoveryAwareChannelN.java:89)
	at com.rabbitmq.client.impl.recovery.AutorecoveringChannel.basicAck(AutorecoveringChannel.java:436)
	at RMQProblemTest$1.handleDelivery(RMQProblemTest.java:94)
	at com.rabbitmq.client.impl.ConsumerDispatcher$5.run(ConsumerDispatcher.java:149)
	at com.rabbitmq.client.impl.ConsumerWorkService$WorkPoolRunnable.run(ConsumerWorkService.java:100)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Consumed so far - 33
Consumed so far - 34
Consumed so far - 34
java.net.SocketException: Broken pipe
	at java.net.SocketOutputStream.socketWrite0(Native Method)
	at java.net.SocketOutputStream.socketWrite(SocketOutputStream.java:109)
	at java.net.SocketOutputStream.write(SocketOutputStream.java:153)
	at java.io.BufferedOutputStream.flushBuffer(BufferedOutputStream.java:82)
	at java.io.BufferedOutputStream.flush(BufferedOutputStream.java:140)
	at java.io.DataOutputStream.flush(DataOutputStream.java:123)
	at com.rabbitmq.client.impl.SocketFrameHandler.flush(SocketFrameHandler.java:177)
	at com.rabbitmq.client.impl.AMQConnection.flush(AMQConnection.java:559)
	at com.rabbitmq.client.impl.AMQCommand.transmit(AMQCommand.java:127)
	at com.rabbitmq.client.impl.AMQChannel.quiescingTransmit(AMQChannel.java:396)
	at com.rabbitmq.client.impl.AMQChannel.transmit(AMQChannel.java:372)
	at com.rabbitmq.client.impl.AMQChannel.transmit(AMQChannel.java:365)
	at com.rabbitmq.client.impl.ChannelN.basicAck(ChannelN.java:1169)
	at com.rabbitmq.client.impl.recovery.RecoveryAwareChannelN.basicAck(RecoveryAwareChannelN.java:89)
	at com.rabbitmq.client.impl.recovery.AutorecoveringChannel.basicAck(AutorecoveringChannel.java:436)
	at RMQProblemTest$1.handleDelivery(RMQProblemTest.java:94)
	at com.rabbitmq.client.impl.ConsumerDispatcher$5.run(ConsumerDispatcher.java:149)
	at com.rabbitmq.client.impl.ConsumerWorkService$WorkPoolRunnable.run(ConsumerWorkService.java:100)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Consumed so far - 34
Consumed so far - 35
Consumed so far - 35
java.net.SocketException: Broken pipe
	at java.net.SocketOutputStream.socketWrite0(Native Method)
	at java.net.SocketOutputStream.socketWrite(SocketOutputStream.java:109)
	at java.net.SocketOutputStream.write(SocketOutputStream.java:153)
	at java.io.BufferedOutputStream.flushBuffer(BufferedOutputStream.java:82)
	at java.io.BufferedOutputStream.flush(BufferedOutputStream.java:140)
	at java.io.DataOutputStream.flush(DataOutputStream.java:123)
	at com.rabbitmq.client.impl.SocketFrameHandler.flush(SocketFrameHandler.java:177)
	at com.rabbitmq.client.impl.AMQConnection.flush(AMQConnection.java:559)
	at com.rabbitmq.client.impl.AMQCommand.transmit(AMQCommand.java:127)
	at com.rabbitmq.client.impl.AMQChannel.quiescingTransmit(AMQChannel.java:396)
	at com.rabbitmq.client.impl.AMQChannel.transmit(AMQChannel.java:372)
	at com.rabbitmq.client.impl.AMQChannel.transmit(AMQChannel.java:365)
	at com.rabbitmq.client.impl.ChannelN.basicAck(ChannelN.java:1169)
	at com.rabbitmq.client.impl.recovery.RecoveryAwareChannelN.basicAck(RecoveryAwareChannelN.java:89)
	at com.rabbitmq.client.impl.recovery.AutorecoveringChannel.basicAck(AutorecoveringChannel.java:436)
	at RMQProblemTest$1.handleDelivery(RMQProblemTest.java:94)
	at com.rabbitmq.client.impl.ConsumerDispatcher$5.run(ConsumerDispatcher.java:149)
	at com.rabbitmq.client.impl.ConsumerWorkService$WorkPoolRunnable.run(ConsumerWorkService.java:100)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Consumed so far - 35
Consumed so far - 36
Consumed so far - 36
java.net.SocketException: Broken pipe
	at java.net.SocketOutputStream.socketWrite0(Native Method)
	at java.net.SocketOutputStream.socketWrite(SocketOutputStream.java:109)
	at java.net.SocketOutputStream.write(SocketOutputStream.java:153)
	at java.io.BufferedOutputStream.flushBuffer(BufferedOutputStream.java:82)
	at java.io.BufferedOutputStream.flush(BufferedOutputStream.java:140)
	at java.io.DataOutputStream.flush(DataOutputStream.java:123)
	at com.rabbitmq.client.impl.SocketFrameHandler.flush(SocketFrameHandler.java:177)
	at com.rabbitmq.client.impl.AMQConnection.flush(AMQConnection.java:559)
	at com.rabbitmq.client.impl.AMQCommand.transmit(AMQCommand.java:127)
	at com.rabbitmq.client.impl.AMQChannel.quiescingTransmit(AMQChannel.java:396)
	at com.rabbitmq.client.impl.AMQChannel.transmit(AMQChannel.java:372)
	at com.rabbitmq.client.impl.AMQChannel.transmit(AMQChannel.java:365)
	at com.rabbitmq.client.impl.ChannelN.basicAck(ChannelN.java:1169)
	at com.rabbitmq.client.impl.recovery.RecoveryAwareChannelN.basicAck(RecoveryAwareChannelN.java:89)
	at com.rabbitmq.client.impl.recovery.AutorecoveringChannel.basicAck(AutorecoveringChannel.java:436)
	at RMQProblemTest$1.handleDelivery(RMQProblemTest.java:94)
	at com.rabbitmq.client.impl.ConsumerDispatcher$5.run(ConsumerDispatcher.java:149)
	at com.rabbitmq.client.impl.ConsumerWorkService$WorkPoolRunnable.run(ConsumerWorkService.java:100)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Consumed so far - 36
Consumed so far - 37
Consumed so far - 37
java.net.SocketException: Broken pipe
	at java.net.SocketOutputStream.socketWrite0(Native Method)
	at java.net.SocketOutputStream.socketWrite(SocketOutputStream.java:109)
	at java.net.SocketOutputStream.write(SocketOutputStream.java:153)
	at java.io.BufferedOutputStream.flushBuffer(BufferedOutputStream.java:82)
	at java.io.BufferedOutputStream.flush(BufferedOutputStream.java:140)
	at java.io.DataOutputStream.flush(DataOutputStream.java:123)
	at com.rabbitmq.client.impl.SocketFrameHandler.flush(SocketFrameHandler.java:177)
	at com.rabbitmq.client.impl.AMQConnection.flush(AMQConnection.java:559)
	at com.rabbitmq.client.impl.AMQCommand.transmit(AMQCommand.java:127)
	at com.rabbitmq.client.impl.AMQChannel.quiescingTransmit(AMQChannel.java:396)
	at com.rabbitmq.client.impl.AMQChannel.transmit(AMQChannel.java:372)
	at com.rabbitmq.client.impl.AMQChannel.transmit(AMQChannel.java:365)
	at com.rabbitmq.client.impl.ChannelN.basicAck(ChannelN.java:1169)
	at com.rabbitmq.client.impl.recovery.RecoveryAwareChannelN.basicAck(RecoveryAwareChannelN.java:89)
	at com.rabbitmq.client.impl.recovery.AutorecoveringChannel.basicAck(AutorecoveringChannel.java:436)
	at RMQProblemTest$1.handleDelivery(RMQProblemTest.java:94)
	at com.rabbitmq.client.impl.ConsumerDispatcher$5.run(ConsumerDispatcher.java:149)
	at com.rabbitmq.client.impl.ConsumerWorkService$WorkPoolRunnable.run(ConsumerWorkService.java:100)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Consumed so far - 37
Consumed so far - 38
Consumed so far - 38
java.net.SocketException: Broken pipe
	at java.net.SocketOutputStream.socketWrite0(Native Method)
	at java.net.SocketOutputStream.socketWrite(SocketOutputStream.java:109)
	at java.net.SocketOutputStream.write(SocketOutputStream.java:153)
	at java.io.BufferedOutputStream.flushBuffer(BufferedOutputStream.java:82)
	at java.io.BufferedOutputStream.flush(BufferedOutputStream.java:140)
	at java.io.DataOutputStream.flush(DataOutputStream.java:123)
	at com.rabbitmq.client.impl.SocketFrameHandler.flush(SocketFrameHandler.java:177)
	at com.rabbitmq.client.impl.AMQConnection.flush(AMQConnection.java:559)
	at com.rabbitmq.client.impl.AMQCommand.transmit(AMQCommand.java:127)
	at com.rabbitmq.client.impl.AMQChannel.quiescingTransmit(AMQChannel.java:396)
	at com.rabbitmq.client.impl.AMQChannel.transmit(AMQChannel.java:372)
	at com.rabbitmq.client.impl.AMQChannel.transmit(AMQChannel.java:365)
	at com.rabbitmq.client.impl.ChannelN.basicAck(ChannelN.java:1169)
	at com.rabbitmq.client.impl.recovery.RecoveryAwareChannelN.basicAck(RecoveryAwareChannelN.java:89)
	at com.rabbitmq.client.impl.recovery.AutorecoveringChannel.basicAck(AutorecoveringChannel.java:436)
	at RMQProblemTest$1.handleDelivery(RMQProblemTest.java:94)
	at com.rabbitmq.client.impl.ConsumerDispatcher$5.run(ConsumerDispatcher.java:149)
	at com.rabbitmq.client.impl.ConsumerWorkService$WorkPoolRunnable.run(ConsumerWorkService.java:100)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Consumed so far - 38
Consumed so far - 39
Consumed so far - 39
java.net.SocketException: Broken pipe
	at java.net.SocketOutputStream.socketWrite0(Native Method)
	at java.net.SocketOutputStream.socketWrite(SocketOutputStream.java:109)
	at java.net.SocketOutputStream.write(SocketOutputStream.java:153)
	at java.io.BufferedOutputStream.flushBuffer(BufferedOutputStream.java:82)
	at java.io.BufferedOutputStream.flush(BufferedOutputStream.java:140)
	at java.io.DataOutputStream.flush(DataOutputStream.java:123)
	at com.rabbitmq.client.impl.SocketFrameHandler.flush(SocketFrameHandler.java:177)
	at com.rabbitmq.client.impl.AMQConnection.flush(AMQConnection.java:559)
	at com.rabbitmq.client.impl.AMQCommand.transmit(AMQCommand.java:127)
	at com.rabbitmq.client.impl.AMQChannel.quiescingTransmit(AMQChannel.java:396)
	at com.rabbitmq.client.impl.AMQChannel.transmit(AMQChannel.java:372)
	at com.rabbitmq.client.impl.AMQChannel.transmit(AMQChannel.java:365)
	at com.rabbitmq.client.impl.ChannelN.basicAck(ChannelN.java:1169)
	at com.rabbitmq.client.impl.recovery.RecoveryAwareChannelN.basicAck(RecoveryAwareChannelN.java:89)
	at com.rabbitmq.client.impl.recovery.AutorecoveringChannel.basicAck(AutorecoveringChannel.java:436)
	at RMQProblemTest$1.handleDelivery(RMQProblemTest.java:94)
	at com.rabbitmq.client.impl.ConsumerDispatcher$5.run(ConsumerDispatcher.java:149)
	at com.rabbitmq.client.impl.ConsumerWorkService$WorkPoolRunnable.run(ConsumerWorkService.java:100)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Consumed so far - 39

…
```

There's evidence of successful client reconnections in the broker log:

```
2018-01-11 11:20:03.889 [info] <0.761.0> accepting AMQP connection <0.761.0> ([fe80::1]:57965 -> [fe80::1]:5672)
2018-01-11 11:20:03.975 [error] <0.761.0> Error on AMQP connection <0.761.0> ([fe80::1]:57965 -> [fe80::1]:5672, state: starting):
PLAIN login refused: user 'guest' can only connect via localhost
2018-01-11 11:20:03.976 [info] <0.761.0> closing AMQP connection <0.761.0> ([fe80::1]:57965 -> [fe80::1]:5672)
2018-01-11 11:20:03.982 [info] <0.766.0> accepting AMQP connection <0.766.0> ([::1]:57966 -> [::1]:5672)
2018-01-11 11:20:04.012 [info] <0.766.0> connection <0.766.0> ([::1]:57966 -> [::1]:5672): user 'guest' authenticated and granted access to vhost '/'
2018-01-11 11:20:04.035 [info] <0.778.0> accepting AMQP connection <0.778.0> ([fe80::1]:57967 -> [fe80::1]:5672)
2018-01-11 11:20:04.049 [error] <0.778.0> Error on AMQP connection <0.778.0> ([fe80::1]:57967 -> [fe80::1]:5672, state: starting):
PLAIN login refused: user 'guest' can only connect via localhost
2018-01-11 11:20:04.050 [info] <0.778.0> closing AMQP connection <0.778.0> ([fe80::1]:57967 -> [fe80::1]:5672)
2018-01-11 11:20:04.057 [info] <0.782.0> accepting AMQP connection <0.782.0> ([::1]:57968 -> [::1]:5672)
2018-01-11 11:20:04.063 [info] <0.782.0> connection <0.782.0> ([::1]:57968 -> [::1]:5672): user 'guest' authenticated and granted access to vhost '/'
2018-01-11 11:20:36.379 [warning] <0.782.0> closing AMQP connection <0.782.0> ([::1]:57968 -> [::1]:5672):
{writer,send_failed,{error,timeout}}
```

**Comment by vincentjames501** [comment]
Thanks @michaelklishin for looking into this! If you are seeing connection recoveries, then that is more than I'm seeing. The moment I see:

```
2018-01-11 07:09:24.930 [warning] <0.548.0> closing AMQP connection <0.548.0> (127.0.0.1:57760 -> 127.0.0.1:5672):
{writer,send_failed,{error,timeout}}
```

I see no additional reconnects to the client (and your log does too?) and Wireshark seems to confirm this as well. I see a two TCP connections established to RabbitMQ from the Java process: the producer connection and consumer connection. Once the TCP buffer is full, RabbitMQ appears to close the connection. No matter how long I wait, I see none of my Shutdown or Recovery listeners fire and Wireshark and RabbitMQ logs seems to indicate no further connections being established.

I haven't looked deeply into the java client library, but I suspect that the reason it keeps consuming messages is because they are simply buffered in memory and we're processing them very slowly (one ever three seconds). If the underlying connection did recover, I suspect that the `basicAck`s would eventually begin to work again, but no matter how long you run the tests they never begin to work.

**Comment by Stephan202** [comment]
This is not my area of expertise, but I can say that we observe a similar issue from time to time, going back to the 4.x series of the client. Symptoms: (a) only happens when a _lot_ of messages are sent and received (single instance) within  a very short time span, (b) only happens when SSL is enabled and (c) recovery doesn't kick in. I'm sorry that I cannot be of more help; we cannot reproduce the issue on demand (which is why so far we haven't filed an issue) and I cannot even say for sure whether what we observe matches @vincentjames501's issue. But it does sound like it.

**Comment by michaelklishin** [comment]
Socket write timeouts can be configured using the `{send_timeout, Integer}` option. They cannot be avoided entirely but perhaps you can have fewer of them. It is also possible to [increase TCP buffer size](http://www.rabbitmq.com/networking.html). Connections then will use more RAM but the probability of running into full TCP buffers and subsequent socket write timeouts will be lower.

Our team does not use issues for investigations or discussions. This one is about listeners not firing even though recovery does happen with the provided test.

**Comment by vincentjames501** [comment]
```This one is about listeners not firing even though recovery does happen with the provided test.```

@michaelklishin , are you positive about this? As I mentioned above I'm not seeing this at all and your log suggested that too from what you posted. Any feedback on my comment https://github.com/rabbitmq/rabbitmq-java-client/issues/341#issuecomment-356932430 ?

**Comment by michaelklishin** [comment]
I re-run the test and here are some more findings.
```

The test does not instruct the client to [recover the topology](http://www.rabbitmq.com/api-guide.html#recovery):

``` java
final ConnectionFactory factory = new ConnectionFactory();
        factory.setAutomaticRecoveryEnabled(true);
        // facory.setTopologyRecoveryEnabled(true); is missing
        factory.setHost("localhost");
```

So even after connection recovery it would not get any more deliveries @vincentjames501 expects.

I see different socket exception variations e.g.

```
2018-01-11 18:49:46.496 [warning] <0.809.0> closing AMQP connection <0.809.0> ([::1]:64100 -> [::1]:5672):
{inet_error,enotconn}
```

According to RabbitMQ's connection state (`rabbitmq_networking:connections/0`), there is one active connection to the node. `netstat` confirms this.

There are two things that I haven't noticed the first time:

 * The test opens two connections.
 * The exception about a broken pipe comes from a `Connection#close` attempt in the `tearDown` method. This suggests the socket is in an interesting state where it cannot be used but the client doesn't seem to be aware of it as no main I/O loop exceptions were raised.

Without inspecting the entire capture and adding a certain amount of debug logging to the client I cannot tell why the client never gets any I/O exceptions on the path that kicks off connection recovery. Once we have more details and can confirm that it can be considered a bug in the library, a new issue will be filed.

Using manual acknowledgements with a limited prefetch avoids the fundamental issue of TCP buffers filling up, so that's the recommendation we have in the meantime.

Thanks again for the provided example.

**Comment by michaelklishin** [comment]
@Stephan202 if what you see only happens with TLS enabled, it could be related to [this rabbitmq-users thread](https://groups.google.com/d/msg/rabbitmq-users/aY_K2jEnAZ0/ddeEBhcTAAAJ).

**Comment by vincentjames501** [comment]
`facory.setTopologyRecoveryEnabled(true)` is enabled by default so adding it is redundant, no?

```java
new ConnectionFactory().isAutomaticRecoveryEnabled(); => true
new ConnectionFactory().isTopologyRecoveryEnabled(); => true
```

I pushed some small changes to the test to make things easier. I now close the producer connection & channel after we're done and name the connections so they are easier to see and visualize. I also added the topology recovery option.

I don't know why this issue keeps getting closed. There is plenty of information here to suggest an issue with the client.

Here is a video showing that there is no reconnection attempts at all made.

[Video](https://www.dropbox.com/s/uid4z7c9qnlxqkz/rmq_issue.mov?dl=0)

**Comment by michaelklishin** [comment]
@vincentjames501 our team does not use GitHub issues for investigations. When we understand what is going, a new issue will be filed. If you have new findings, feel free to start a rabbitmq-users thread.

**Comment by michaelklishin** [comment]
For those who are affected: the workaround is to use [manual acknowledgements with a prefetch](http://www.rabbitmq.com/confirms.html).

**Comment by michaelklishin** [comment]
@lukebakken convinced me we have enough details to work with here.

**Comment by Stephan202** [comment]
Thanks for the pointer @michaelklishin, I'll review that thread. Indeed, we are already using manual acknowledgements with a prefetch (of 20 messages), so it's unlikely then that issue we're observing matches the topic of this thread.

**Comment by acogoluegnes** [comment]
I imported the project but couldn't reproduce the issue. I got the following twice:

```
Consumed so far - 78
Consumed so far - 79
Consumed so far - 79
Consumed so far - 79

java.lang.AssertionError: We consume our messages within the specified time. We consumed 80 messages.
```

My environment:
* RabbitMQ 3.6.12 and 3.7.0
* Erlang 20.0.1
* Java 1.8.0_152
* Java Client 4.4.1
* Ubuntu 14.04, 4.4.0-109-generic

**Comment by vincentjames501** [comment]
@acogoluegnes , FWIW I can't get it to happen when running RabbitMQ inside the official RabbitMQ docker container. It is easily reproducible on an OSX machine running the official RabbitMQ binaries directly on the host (OSX 10.13.2). I forgot to note that in the original bug report. We DO see the exact same behavior in our AWS environment, however, so I don't think it is a Linux vs OSX thing either. I *think* the reason it works in Docker for me is because it is substantially slower when using persistent messages than it is when I run it natively but it could be a lower level networking difference as well (though seems unlikely?).

**Comment by michaelklishin** [comment]
I can reproduce it on a 4 year old Mac easily. I suspect the rate of deliveries has to be at least so high for TCP window to be exhausted completely.

**Comment by vincentjames501** [comment]
@acogoluegnes , you may be able to get it to happen by disabling persistent messages to increase the rate of delivery? Not entirely sure but I created that test to be a boiled down version of exactly what we are doing in production so I didn't tinker with many more settings.

**Comment by acogoluegnes** [comment]
Reproduced as-is on my 2015 MBP:

```
Consumed so far - 0
Consumed so far - 0
Consumed so far - 0
Consumed so far - 0
Consumed so far - 1
...
Consumed so far - 10
Consumed so far - 10
Consumed so far - 10
Consumed so far - 11
Consumed so far - 11
Problem acking message Broken pipe
Consumed so far - 11
Consumed so far - 12
Consumed so far - 12
Problem acking message Broken pipe
Consumed so far - 12
Consumed so far - 13
Consumed so far - 13
Problem acking message Broken pipe
```

Reproduced on Linux by setting the number of messages to 100 000:
```
Consumed so far - 0
Consumed so far - 0
Consumed so far - 0
Consumed so far - 0
Consumed so far - 1
Consumed so far - 1
Consumed so far - 1
Consumed so far - 9
Consumed so far - 9
Consumed so far - 9
Consumed so far - 10
Consumed so far - 10
Consumed so far - 10
Consumed so far - 11
Consumed so far - 11
Problem acking message Connection reset
Consumed so far - 11
Consumed so far - 12
Consumed so far - 12
Problem acking message Broken pipe (Write failed)
Consumed so far - 12
Consumed so far - 13
Consumed so far - 13
Problem acking message Broken pipe (Write failed)
```

**Comment by acogoluegnes** [comment]
A thread dump reveals the reading thread is stuck waiting for a delivery to be put on the work queue. Unfortunately, this operation has not timeout. Needs more digging to know why the consumption of deliveries doesn't unblock the working queue.

**Comment by acogoluegnes** [comment]
The consumption does unblock the working queue, but more deliveries come in. They must be the remaining of the saturated TCP buffer. Automatic connection recovery doesn't kick in as it's started only on the reading side. The reading thread will finally notice something is wrong once it has finished reading the TCP buffer, so this can take a while.

There's nothing much we can do to detect the connection is dead, considering this is a typical case of the broker overflowing a slow client and that's why QoS has been implemented for.

We can make the work queue capacity configurable (the default is 1000 per channel). Making it bigger wouldn't hurt in cases like this one. WDYT @vincentjames501 @michaelklishin?

**Comment by vincentjames501** [comment]
I haven't spent much time analyzing the internals of the library to give a thoughtful response, however, there must be some hook that the underlying TCP connection is disconnected, no? A packet capture clearly shows the connection is terminated, shouldn't the consuming thread notice this and kick off a recovery (and stop delivering more work to the subscribers)? 

Would you mind explaining (because of my ignorance) why this is different than a typical connection failure? If I just terminate the connection at RabbitMQ connection recovery initiates immediately even if there are unacked messages.

**Comment by michaelklishin** [comment]
@vincentjames501 this feature has been around for years, no need to assume it doesn't handle most obvious failure scenarios.

If we ignore NIO for moment, every connection has an I/O loop thread which [handles all exceptions and starts a shutdown procedure](https://github.com/rabbitmq/rabbitmq-java-client/blob/master/src/main/java/com/rabbitmq/client/impl/AMQConnection.java#L577). Connections with enabled automatic recovery [register a shutdown hook](https://github.com/rabbitmq/rabbitmq-java-client/blob/master/src/main/java/com/rabbitmq/client/impl/recovery/AutorecoveringConnection.java#L444) that [kicks off the recovery](https://github.com/rabbitmq/rabbitmq-java-client/blob/master/src/main/java/com/rabbitmq/client/impl/recovery/AutorecoveringConnection.java#L507).

When TCP window is saturated, there is no I/O exception thrown by the JDK even after the server detects missed heartbeats and closes the connection. If knew why this would likely have been addressed already.

**Comment by michaelklishin** [comment]
@acogoluegnes seems to have a hypothesis, we are discussing it at the moment.

**Comment by michaelklishin** [comment]
@acogoluegnes I think we should try making the interval lower, or rather, in line with the heartbeat mechanism. Having a consumer work queue timeout would be an improvement but it wouldn't address the root cause here unless I misunderstand the above explanation.

**Comment by michaelklishin** [comment]
This reminds me of a similar (but not the same) scenario that took us a long time to handle well in the server, see https://github.com/rabbitmq/rabbitmq-common/pull/31. Socket implementations can behave differently when it comes to event/error reporting on sockets with saturated buffers (windows).
In the server's case we made the heartbeat monitor take more responsibility and force error condition detection in the connection process.

It should be possible to do something similar with JDK sockets, although it can be hacky as hell.

**Comment by acogoluegnes** [comment]
@vincentjames501 Believe it or not, but there's no hook to know the TCP connection is terminated with Java's `Socket` API.

As @michaelklishin suggested, we should be able to notify the reading thread that the connection is gone from the heartbeat sender or from any writing operation that has failed due a network problem. The only way is to interrupt the reading thread. Interrupting a thread in Java doesn't carry much information, but this should be enough to know when to trigger recovery.

**Comment by michaelklishin** [comment]
With a version I'm testing (#349), I observe a successful recovery:

```
2018-02-14 17:20:11.863 [info] <0.1005.0> accepting AMQP connection <0.1005.0> ([::1]:64548 -> [::1]:5672)
2018-02-14 17:20:12.085 [info] <0.1005.0> Connection <0.1005.0> ([::1]:64548 -> [::1]:5672) has a client-provided name: Producer Connection
2018-02-14 17:20:12.096 [info] <0.1005.0> connection <0.1005.0> ([::1]:64548 -> [::1]:5672 - Producer Connection): user 'guest' authenticated and granted access to vhost '/'
2018-02-14 17:20:12.112 [info] <0.1018.0> accepting AMQP connection <0.1018.0> ([::1]:64549 -> [::1]:5672)
2018-02-14 17:20:12.114 [info] <0.1018.0> Connection <0.1018.0> ([::1]:64549 -> [::1]:5672) has a client-provided name: Consuming Connection
2018-02-14 17:20:12.116 [info] <0.1018.0> connection <0.1018.0> ([::1]:64549 -> [::1]:5672 - Consuming Connection): user 'guest' authenticated and granted access to vhost '/'
2018-02-14 17:20:13.927 [info] <0.1005.0> closing AMQP connection <0.1005.0> ([::1]:64548 -> [::1]:5672 - Producer Connection, vhost: '/', user: 'guest')
2018-02-14 17:20:43.553 [warning] <0.1018.0> closing AMQP connection <0.1018.0> ([::1]:64549 -> [::1]:5672 - Consuming Connection):
{writer,send_failed,{error,timeout}}
2018-02-14 17:21:11.298 [info] <0.1067.0> accepting AMQP connection <0.1067.0> ([::1]:64563 -> [::1]:5672)
2018-02-14 17:21:11.303 [info] <0.1067.0> Connection <0.1067.0> ([::1]:64563 -> [::1]:5672) has a client-provided name: Consuming Connection
2018-02-14 17:21:11.305 [info] <0.1067.0> connection <0.1067.0> ([::1]:64563 -> [::1]:5672 - Consuming Connection): user 'guest' authenticated and granted access to vhost '/'
```

```
Consumed so far - 18
Consumed so far - 19
Recovery Started for Consuming Connection amqp://guest@0:0:0:0:0:0:0:1:5672/
Recovery Started for Consuming Channel AMQChannel(amqp://guest@0:0:0:0:0:0:0:1:5672/,1)
Recovery finished for Consuming Channel AMQChannel(amqp://guest@0:0:0:0:0:0:0:1:5672/,1)
Recovery finished for Consuming Connection amqp://guest@0:0:0:0:0:0:0:1:5672/
Consumed so far - 19
Consumed so far - 19
Consumed so far - 20
Consumed so far - 20
```
